SFDO-Tooling/MetaDeploy
# Expose Admin API to non-admin users

## Changes
* The `/admin/rest` API is now located under `/tenant/rest`, and can be optionally overriden with the `TENANT_AREA_PREFIX` environment variable.
* This API is available to tenant admins who have a valid access token for the tenant they are calling out to, and CRUD permissions for the models they are operating on. Session authentication is not permitted on these API endpoints.
* The django admin UI is still located under DJANGO_ADMIN_URL. Django-rq urls have also been moved under the DJANGO_ADMIN_URL prefix.
* Altered the multi-tenancy middleware to work with the new url prefix structure.

## Critical Changes
* Existing MetaDeploy services present in `cci` will no longer work. They will need to be re-connected using the new endpoint (`/tenant/rest`).

## QA Performed
* Registered a MetaDeploy service on cci using a token associated to tenant admin (non-super) user. Successfully able to run the `metadeploy_publish` task against EDA v1.120.

1. Logged in as a super user, ensure that a Site record exists with a domain of `localhost:8080`.
2. Create a "Tenant Admins" Permission group. Assign CRUD permissions for all models under the API application.
3. Login with devhub credentials and Django AllAuth will provision a new user.
4. In an incognito window, login as a super user, locate the new user that corresponds to your devhub user, and assign them the "Tenant Admins" group permissions.
5. Create a new token, and select the `localhost:8080` Site and devhub user for the lookups on the token.
6. Connect a new metadeploy service via `cci service connect metadeploy local`. For the domain enter `http://localhost:8080/tenant/admin`, copy and paste the token value from the django admin when prompted to enter it.
7. Clone the EDA repository and `cd` into it.
8. Run `cci task run metadeploy_publish --plan install --tag rel.1.120`
9. Verify that the EDA 1.120 Install plan was successfully uploaded via the django admin UI.

* Logged in as a tenant admin, and verified that `/tenant/rest/` endpoints are not accessible with a valid session.
* Verified that super users are still able to toggle between tenants via the picklist present in the Admin UI.
* The naming change from ADMIN_AREA_PREFIX to TENANT_AREA_PREFIX will not impact any of our apps as this environment variable is not currently in use.
